cilium/cilium
# Services support for external workloads

Refactor operator's services watcher to be usable also from `clustermesh-apiserver`. Make Cilium agent sync services from kvstore in `join-cluster` mode. Use `cilium service` to locate `kube-dns` and overwrite VM's `/etc/resolv.conf` with it.

None of this should affect any other functionality, so this should be safe to merge.

In `pkg/k8s/service_cache.go`, function `correlateEndpoints()` a check for `option.Config.ClusterName` is removed. This is required as cluster services synced from the k8s cluster will have the same cluster name as the external endpoint has, as it is conceptually in the same cluster as the k8s cluster it joined. Clustermesh call paths have the same check so it is still not possible for a clustermesh to re-introduce backends for it's own cluster via external endpoints.

'experimental-install.yaml' is changed to not configure support for external workloads, so an explicit helm option is required. Also, a new helm value `clustermesh.useAPIServer` is added and `clustermesh-apiserver` is only deployed if that or `externalWorkloads.enabled` is `true`, both default to `false`. This avoids surprising existing users of clustermesh with a new deployment.

Scripts are added to `contrib/k8s` to extract TLS certs as well as to install Cilium agent onto a VM with docker. Service access requires `--enable-host-reachable-services`, which does not work on older kernels. We could soften this requirement by not requiring that service access from VM is usable. Currently kube-dns is accessed via it's ClusterIP service, though.

TODO:
- [ ] CI test(s)
- [x] GSG

For later PRs:
- [ ] Handle headless services (== no frontends)
- [ ] test with IPv6 instead of IPv4
- [ ] test with direct routing modes, remove the need for IPAM setup in this case
- [ ] check if tunneling can be done without any IPAM (i.e., w/o creating `cilium_host` device)
- [ ] pass policy from cluster to the VMs
- [ ] enforce policy in VMs

Fixes: #13666

Test setup consists of two VMs, one running with k8s (`k8s1`), one running without k8s (`runtime`). Setup steps:

1. `cd cilium/test`
2. Start `k8s1`:
`VM_CPUS=6 VM_MEMORY=5120 K8S_NODES=1 ./vagrant-local-start.sh`
 - This may take a long time for the 1st run, as a vagrant box is packaged (into `test/.vagrant/k8s-box-package.box`). Subsequent runs will be faster. This packaging removes the need to pull test images for later runs of `./vagrant-local-start.sh`.
3. `vagrant ssh k8s1-1.19`
4. `cd ~/go/src/github.com/cilium/cilium`
5. `DOCKER_BUILDKIT=1 make docker-cilium-image docker-operator-generic-image docker-clustermesh-apiserver-image`
6. `helm install cilium install/kubernetes/cilium --namespace kube-system --set externalWorkloads.enabled=true --set image.pullPolicy=IfNotPresent --set operator.image.pullPolicy=IfNotPresent --set preflight.image.pullPolicy=IfNotPresent --set clustermesh.apiserver.image.pullPolicy=IfNotPresent --set debug.enabled=true`
7. Apply the CiliumExternalWorkload CRD for VM named `runtime`: `kubectl apply -f clustermesh-apiserver/test/runtime.yaml`
8. Extract the TLS client keys to be used by the VM that is started up in the next step:
`contrib/k8s/extract-external-workload-certs.sh`
9. In another terminal, change to `test` directory in your Cilium repo and start the `runtime` VM:
`PROVISION_EXTERNAL_WORKLOAD=1 PRELOAD_VM=false KERNEL=419 VM_CPUS=6 VM_MEMORY=4096 ./vagrant-local-start-runtime.sh`
10. `vagrant ssh runtime`
11. Check that etcd has connected successfully, observe the first line of:
`cilium status`
You should see something like:
`etcd: 1/1 connected, lease-ID=7c02748328e75f57, lock lease-ID=7c02748328e75f59, has-quorum=true: 192.168.36.11:32379 - 3.4.13 (Leader)`
12. Check that cluster DNS works: `nslookup clustermesh-apiserver.kube-system.svc.cluster.local`
13. Ping the backend IP of `clustermesh-apiserver` service:
`ping $(cilium service list get -o jsonpath='{[?(@.spec.flags.name=="clustermesh-apiserver")].spec.backend-addresses[0].ip}')`
14. The ping should keep running also when the following CCNP is applied in `k8s1` (e.g., `kubectl apply -n kube-system -f clustermesh-apiserver/test/testccnp.yaml`):
```
apiVersion: cilium.io/v2
kind: CiliumClusterwideNetworkPolicy
metadata:
  name: test-ccnp
  namespace: kube-system
spec:
  endpointSelector:
    matchLabels:
      k8s-app: clustermesh-apiserver
  ingress:
  - fromEndpoints:
    - matchLabels:
        io.kubernetes.pod.name: runtime
  - toPorts:
    - ports:
      - port: "2379"
        protocol: TCP
```
15. The ping should stop if you delete these lines from the policy and re-apply:
```
  - fromEndpoints:
    - matchLabels:
        io.kubernetes.pod.name: runtime
```
16. `kubectl get cep` should show an entry for `runtime` with `IDENTITY ID` and `IPV4` set.
17. `kubectl get cew` should show an entry for `runtime` with matching `CILIUM ID` and `IP`.
